File: algorithms/416. Partition Equal Subset Sum/solution.py
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        # we use a big int, to indicate whether a sum
        # has been appeared
        s, bit = 0, 1
        for n in nums:
            s += n
            bit |= bit << n
        return not (s & 1) and bool(bit & (1 << (s / 2)))

File: algorithms/416. Partition Equal Subset Sum/solution.cc
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        bitset<10001> bits(1);
        for ( int n : nums ) bits |= bits << n;
        return ! (sum & 1) && bits[sum / 2];
    }
};
